### PR TITLE
Filter XCOM by key when calculating map lengths

### LINKED ISSUE #23792: Dynamic task mapping creates too many mapped instances when task pushed non-default XCom
### Apache Airflow version

2.3.0 (latest released)

### What happened

Excess tasks are created when using dynamic task mapping with KubernetesPodOperator, but only in certain cases which I do not understand. I have a simple working example of this where the flow is:
- One task that returns a list XCom (list of lists, since I'm partial-ing to `KubernetesPodOperator`'s `arguments`) of length 3. This looks like `[["a"], ["b"], ["c"]]`
- A `partial` from this, which is expanded on the above's result. Each resulting task has an XCom of a single element list that looks like `["d"]`. We expect the `expand` to result in 3 tasks, which it does. So far so good. Why doesn't the issue occur at this stage? No clue.
- A `partial` from the above. We expect 3 tasks in this final stage, but get 9. 3 succeed and 6 fail consistently. This 3x rule scales to as many tasks as you define in step 2 (e.g. 2 tasks in step 2 -> 6 tasks in step 3, where 4 fail)

image

If I recreate this using the TaskFlow API with `PythonOperator`s, I get the expected result of 1 task -> 3 tasks -> 3 tasks

image

Futhermore, if I attempt to look at the `Rendered Template` of the failed tasks in the `KubernetesPodOperator` implementation (first image), I consistently get `Error rendering template` and all the fields are `None`. The succeeded tasks look normal.

image

Since the `Rendered Template` view fails to load, I can't confirm what is actually getting provided to these failing tasks' `argument` parameter. If there's a way I can query the meta database to see this, I'd be glad to if given instruction. 

### What you think should happen instead

I think this has to do with how XComs are specially handled with the `KubernetesPodOperator`. If we look at the XComs tab of the upstream task (`some-task-2` in the above images), we see that the return value specifies `pod_name` and `pod_namespace` along with `return_value`.

image

Whereas in the `t2` task of the TaskFlow version, it only contains `return_value`. 

image

I haven't dug through the code to verify, but I have a strong feeling these extra values `pod_name` and `pod_namespace` are being used to generate the `OperatorPartial`/`MappedOperator` as well when they shouldn't be.

### How to reproduce

Run this DAG in a k8s context:

```
from datetime import datetime
from airflow import XComArg
from airflow.models import DAG
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator


def make_operator(
    **kwargs
):
    return KubernetesPodOperator(
        **{
            'get_logs': True,
            'in_cluster': True,
            'is_delete_operator_pod': True,
            'namespace': 'default',
            'startup_timeout_seconds': 600,
            **kwargs,
        }
    )


def make_partial_operator(
    **kwargs
):
    return KubernetesPodOperator.partial(
        **{
            'get_logs': True,
            'in_cluster': True,
            'is_delete_operator_pod': True,
            'namespace': 'default',
            'startup_timeout_seconds': 600,
            **kwargs,
        }
    )


with DAG(dag_id='test-pod-xcoms',
         schedule_interval=None,
         start_date=datetime(2020, 1, 1),
         max_active_tasks=20) as dag:

    op1 = make_operator(
        cmds=['python3', '-c' 'import json;f=open("/airflow/xcom/return.json", "w");f.write(json.dumps([["a"], ["b"], ["c"]]))'],
        image='python:3.9-alpine',
        name='airflow-private-image-pod-1',
        task_id='some-task-1',
        do_xcom_push=True
    )

    op2 = make_partial_operator(
        cmds=['python3', '-c' 'import json;f=open("/airflow/xcom/return.json", "w");f.write(json.dumps(["d"]))'],
        image='python:3.9-alpine',
        name='airflow-private-image-pod-2',
        task_id='some-task-2',
        do_xcom_push=True
    )

    op3 = make_partial_operator(
        cmds=['echo', 'helloworld'],
        image='alpine:latest',
        name='airflow-private-image-pod-3',
        task_id='some-task-3',
    )

    op3.expand(arguments=XComArg(op2.expand(arguments=XComArg(op1))))
```

For the TaskFlow version of this that works, run this DAG (doesn't have to be k8s context):

```
from datetime import datetime

from airflow.decorators import task
from airflow.models import DAG, Variable


@task
def t1():
    return [[1], [2], [3]]


@task
def t2(val):
    return val


@task
def t3(val):
    print(val)


with DAG(dag_id='test-mapping',
         schedule_interval=None,
         start_date=datetime(2020, 1, 1)) as dag:

    t3.partial().expand(val=t2.partial().expand(val=t1()))
```

### Operating System

MacOS 11.6.5

### Versions of Apache Airflow Providers

Relevant:

```
apache-airflow-providers-cncf-kubernetes==4.0.1
```

Full:

```
apache-airflow-providers-amazon==3.3.0
apache-airflow-providers-celery==2.1.4
apache-airflow-providers-cncf-kubernetes==4.0.1
apache-airflow-providers-docker==2.6.0
apache-airflow-providers-elasticsearch==3.0.3
apache-airflow-providers-ftp==2.1.2
apache-airflow-providers-google==6.8.0
apache-airflow-providers-grpc==2.0.4
apache-airflow-providers-hashicorp==2.2.0
apache-airflow-providers-http==2.1.2
apache-airflow-providers-imap==2.2.3
apache-airflow-providers-microsoft-azure==3.8.0
apache-airflow-providers-mysql==2.2.3
apache-airflow-providers-odbc==2.0.4
apache-airflow-providers-postgres==4.1.0
apache-airflow-providers-redis==2.0.4
apache-airflow-providers-sendgrid==2.0.4
apache-airflow-providers-sftp==2.6.0
apache-airflow-providers-slack==4.2.3
apache-airflow-providers-sqlite==2.1.3
apache-airflow-providers-ssh==2.4.3
```

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

Docker (Docker Desktop)

- Server Version: 20.10.13
- API Version: 1.41
- Builder: 2

Kubernetes (Docker Desktop)

- Env: docker-desktop
- Context: docker-desktop
- Cluster Name: docker-desktop
- Namespace: default
- Container Runtime: docker
- Version: v1.22.5

Helm:

- version.BuildInfo{Version:"v3.6.3", GitCommit:"d506314abfb5d21419df8c7e7e68012379db2354", GitTreeState:"dirty", GoVersion:"go1.16.5"}

### Anything else

_No response_

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #24487: Dynamic mapping over KubernetesPodOperator results produces triplicate child tasks
### Apache Airflow version

2.3.2 (latest released)

### What happened

Attempting to use [dynamic task mapping](https://airflow.apache.org/docs/apache-airflow/2.3.0/concepts/dynamic-task-mapping.html#mapping-over-result-of-classic-operators) on the results of a `KubernetesPodOperator` (or `GKEStartPodOperator`) produces 3x as many downstream task instances as it should. Two-thirds of the downstream tasks fail more or less instantly.

### What you think should happen instead

The problem is that the number of downstream tasks is calculated by counting XCOMs associated with the upstream task, assuming that each `task_id` has a single XCOM:
https://github.com/apache/airflow/blob/fe5e689adfe3b2f9bcc37d3975ae1aea9b55e28a/airflow/models/mappedoperator.py#L606-L615

However the `KubernetesPodOperator` pushes two XCOMs in its `.execute()` method:

https://github.com/apache/airflow/blob/fe5e689adfe3b2f9bcc37d3975ae1aea9b55e28a/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py#L425-L426

So the number of downstream tasks ends up being 3x what it should.

### How to reproduce

Reproducing the behavior requires access to a kubernetes cluster, but in psedo-code, a dag like this should demonstrate the behavior:
```
with DAG(...) as dag:

    # produces an output list with N elements
    first_pod = GKEStartPodOperator(..., do_xcom_push=True)

    # produces 1 output per input, so N task instances are created each with a single output
    second_pod = GKEStartPodOperator.partial(..., do_xcom_push=True).expand(id=XComArg(first_pod))

    # should have N task instances created, but actually gets 3N task instances created
    third_pod = GKEStartPodOperator.partial(..., do_xcom_push=True).expand(id=XComArg(second_pod))
```

### Operating System

macOS 12.4

### Versions of Apache Airflow Providers

apache-airflow-providers-cncf-kubernetes==4.1.0
apache-airflow-providers-google==8.0.0

### Deployment

Virtualenv installation

### Deployment details

_No response_

### Anything else

When I edit `mappedoperator.py` in my local deployment to filter on the XCOM key things behave as expected:
```
        # Collect lengths from mapped upstreams.
        xcom_query = (
            session.query(XCom.task_id, func.count(XCom.map_index))
            .group_by(XCom.task_id)
            .filter(
                XCom.dag_id == self.dag_id,
                XCom.run_id == run_id,
                XCom.key == 'return_value',             <------- added this line
                XCom.task_id.in_(mapped_dep_keys),
                XCom.map_index >= 0,
            )
        )
```

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by hinnefe2**
MappedOperator determines how many downstream task instances to create
by counting XCOMs associated with upstream mapped tasks. The current
implementation assumes each upstream mapped task has a single XCOM. When
upstream tasks push additional XCOMs beyond the default one then
MappedOperator will generate too many downstream task instances. This
change updates the XCOM counting query to only consider XCOMs with a key
matching the `XCOM_RETURN_KEY` constant.

close #24487
close #23792

### CODE REVIEW

**Review by ashb — COMMENTED** [review]
Change looks good - can you add to the tests please?

**Comment by jedcunningham** [comment]
@hinnefe2, gentle ping. It'd be great to get this into 2.3.3. That test file is the right spot. If you have further questions/get stuck, feel free to ping us in #development on Slack!

**Comment by hinnefe2** [comment]
Thanks for the nudge. I'm having a little trouble getting the test to work as intended but I'll sort that out on Slack.

File: tests/test_utils/mock_operators.py [hunk 1]
```diff
@@ -161,3 +161,17 @@ def __init__(self, **kwargs):
...
+    Operator which pushes an extra XCOM value along with the default XCOM_RETURN_KEY
+    """
+
+    def __init__(self, return_value, **kwargs):
+        super().__init__(do_xcom_push=True, **kwargs)
+        self.return_value = return_value
+
+    def execute(self, context: Context):
+        context['task_instance'].xcom_push(key='extra_key', value="extra_value")
+        return self.return_value
```
Comment (uranusjr) [inline tests/test_utils/mock_operators.py]: Since there’s only one test that needs this class, I’d put it in the test instead.

**Review by uranusjr — COMMENTED** [review]
One styling nitpick, the logic looks good to me.

**Review by uranusjr — APPROVED** [review]
I pushed a couple of commits to resolve conflicts and update the test.

**Comment by uranusjr** [comment]
I’m going to make the call since this is straightfoward enough.

**Comment by hinnefe2** [comment]
🎉 Thanks for the help getting this across the finish line!

**Pull request merged**